compiler/ilgen/VirtualMachineState.hpp

```diff
@@ -0,0 +1,123 @@
+ *
+ * VirtualMachineState is an interface that language compilers will implement
+ * to define all the aspects of the virtual machine state that the JIT compiler will
+ * simulate (operand stacks, virtual registers like "pc", "sp", etc.). The interface
+ * has four functions: 1) Commit() to cause the simulated state to be written to the
+ * actual virtual machine state, 2) Reload() to set the simulated machine state based
+ * on the current virtual machine state, 3) MakeCopy() to make an identical copy of
+ * the curent virtual machine state object, and 4) MergeInto() to perform the actions
+ * needed for the current state to be merged with a second virtual machine state.
+ *
+ * ##Usage
+ *
+ * Commit() is typically called when transitioning from compiled code to the interpreter.
+ *
+ * Reload() is typically called on the transition back from the interpreter to compiled
+ * code.
+ *
+ * MakeCopy() is typically called when control flow edges are created: the current vm state
+ * must also become the initial vm state at the target of the control flow edge and be
+ * able to evolve independently from the current vm state.
+ *
+ * MergeInto() is typically called when one compiled code path needs to merge into a second
+ * compiled code path (think the bottom of an if-then-else control flow diamond). At the
+ * merge point, the locations of all aspects of the simulated machine state must be
+ * identical so that the code below the merge point will compute correct results.
+ * MergeInto() typically causes the current simulated machine state to be written to the
+ * locations that were used by an existing simulated machine state. The "other" object
+ * passed to MergeInto() must have the exact same shape (number of components and each
+ * component must also match in its type).
+ *
+ * Language compilers should extend this base class, add instance variables for all
+ * necessary virtual machine state variables (using classes like VirtualMachineRegister
+ * and VirtualMachineOperandStack) and then implement Commit(), Reload(), MakeCopy(),
+ * and MergeInto() to call Commit(), Reload(), MakeCopy(), and MergeInto(), respectively, on
+ * each individual instance variable. It feels like boilerplate, but this approach makes
+ * it easy for the compiler to reference individual virtual machine state variables and
+ * at the same time permits complete flexibility in how the VM-wide Commit(), Reload(),
+ * MakeCopy(), and MergeInto() operations can be implemented.
+ */
+
+class VirtualMachineState
+   {
+   public:
+
+   /**
+    * @brief Cause all simulated aspects of the virtual machine state to become real.
+    * @param b builder object where the operations will be added to change the virtual machine state.
+    *
+    * The builder object b is assumed to be along a control flow path transitioning
+    * from compiled code to the interpreter. Base implementation does nothing.
+    */
+   virtual void Commit(TR::IlBuilder *b) { }
```

Comment:
Hm... I see your point about `MergeInto`. However, I can't think of a case where `Commit` and `Reload` don't need to be re-implemented. If there are such cases, I would think they are rather rare. So, to help enforce the idea that this is really an interface, I would prefer if these were pure virtual.

That said, since this is more of a conceptual issue than a functional issue, I'm okay with leaving it as is.